From d2b76cc0d9e2756eadac9f5e0f6c343c4305342a Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:17:45 -0400
Subject: [net] netfilter: nf_tables: fix bogus rulenum after goto action

Message-id: <20141021161745.22888.92114.stgit@localhost.localdomain>
Patchwork-id: 97943
O-Subject: [RHEL7 net PATCH 046/124] netfilter: nf_tables: fix bogus rulenum after goto action
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: f7e7e39b21c285ad73a62fac0736191b8d830704

 commit f7e7e39b21c285ad73a62fac0736191b8d830704
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Sat May 10 18:46:02 2014 +0200

    netfilter: nf_tables: fix bogus rulenum after goto action

    After returning from the chain that we just went to with no matchings,
    we get a bogus rule number in the trace. To fix this, we would need
    to iterate over the list of remaining rules in the chain to update the
    rule number counter.

    Patrick suggested to set this to the maximum value since the default
    base chain policy is the very last action when the processing the base
    chain is over.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_core.c b/net/netfilter/nf_tables_core.c
index f3e3892..711c7f6 100644
--- a/net/netfilter/nf_tables_core.c
+++ b/net/netfilter/nf_tables_core.c
@@ -202,7 +202,7 @@ next_rule:
  }
 
  if (unlikely(pkt->skb->nf_trace))
-  nft_trace_packet(pkt, basechain, ++rulenum, NFT_TRACE_POLICY);
+  nft_trace_packet(pkt, basechain, -1, NFT_TRACE_POLICY);
 
  rcu_read_lock_bh();
  stats = rcu_dereference(nft_base_chain(basechain)->stats);
-- 
1.7.1